### DISCUSSION TITLE
Add support for conditional logical compound assignments

### DISCUSSION THREAD

**Discussion by synercoder** [dotnet/csharplang#6827] [Language Ideas] [answered]
### TL;DR: 
I would like the `&&=` and `||=` operators to become a thing in C#.

---
### Introduction
According to the [Boolean logical operators](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators) docs page; In case of the "Logical AND operator &" and the "Logical OR operator |", both operands are evaluated in the expression:

[Logical OR operator |](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#logical-or-operator-)
> The | operator evaluates both operands even if the left-hand operand evaluates to true, so that the operation result is true regardless of the value of the right-hand operand.

[Logical AND operator &](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#logical-and-operator-)
> The & operator evaluates both operands even if the left-hand operand evaluates to false, so that the operation result is false regardless of the value of the right-hand operand.

Both operations might be handy in some use cases, but most people will use the [Conditional logical AND operator &&](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#logical-or-operator-) and the [ConditionalLogical OR operator ||](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#logical-or-operator-). Which both state:

>The conditional logical AND operator &&, also known as the "short-circuiting" logical AND operator, computes the logical AND of its operands. The result of x && y is true if both x and y evaluate to true. Otherwise, the result is false. **If x evaluates to false, y isn't evaluated.**

>The conditional logical OR operator ||, also known as the "short-circuiting" logical OR operator, computes the logical OR of its operands. The result of x || y is true if either x or y evaluates to true. Otherwise, the result is false. **If x evaluates to true, y isn't evaluated.**

While the first two operators do have compound assignment versions of them `|=` and `&=`. The later operators do not. [The documation](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#compound-assignment) indicates as much:

> The conditional logical operators && and || don't support compound assignment.

### Example use case

The following code is an example where I can use the conditional logic operator to make sure expensive calls are not executed if the boolean was already true:
```c#
public async Task<bool> IsUserAdmin(User user)
{
    var bool isAdmin = _checkIfUserIsLocalAdmin(user);           // Quick check
    isAdmin = isAdmin || await _checkIfUserIsGroupAdmin(user);   // Uses an "expensive" network call   
    isAdmin = isAdmin || await _checkIfUserIsCompanyAdmin(user); // Uses an "expensive" network call

    return isAdmin;
}
```

But why not make the conditional compound assignment operators also available? So the code could be rewritten as:

```C#
public async Task<bool> IsUserAdmin(User user)
{
    var bool isAdmin = _checkIfUserIsLocalAdmin();  // Quick check
    isAdmin ||= await _checkIfUserIsGroupAdmin();   // Uses an "expensive" network call   
    isAdmin ||= await _checkIfUserIsCompanyAdmin(); // Uses an "expensive" network call

    return isAdmin;
}
```

When writing this code, you could of course now use the existing `|=` operator, but it will evaluate both sides of the expression, and thus always make the "expensive" calls.

This of course is just one example where those operators might be useful. 

---

### Historic discussion

The question was asked a long time ago on [stackoverflow](https://stackoverflow.com/questions/6346001/why-are-there-no-or-operators-in-c). This was before C# 5.0 was even released, Since the language has evolved much over the last decade, I do not know if many of the below points would still hold up, I added them below for context and my view on them as well.

[Eric Lippert](https://github.com/ericlippert) answered that in his opinion those operators would be silly because there are alternatives that would work better in situations, and only a few people would use them.

>The || and && operators do not have a compound form because frankly, they're a bit silly. Under what circumstances would you want to say
>
>`b1 ||= b2;`
>`b1 &&= b2;`
>such that the right hand side is not evaluated if the left hand side does not change? It seems like only a few people would actually use this feature, so why put it in?

In the comments he answered a few use cases and what code you could use instead. He also pointed out that in some cases it might even make your code slower:

>Moreover: the point of &&= over &=, which you already have, would be to avoid the cost of the right hand side when the left hand side is false. Does that really avoid a cost in your case? You are trading a test and a branch in exchange for skipping an array read -- cheap -- and a logical and -- super cheap! This would be in many cases larger, slower code. Larger code means more frequent cache misses, more branches means more opportunities for the branch predictor to get it wrong. More basic blocks means the jitter has more work to do. Not worth it. 

---
### My (counter) points
While the original answer to the question was posted by mr. Lippert in 2011, and his comments in 2016, I would also like to start the discussion here. Because I do not fully agree with mr. Lippert here. While mr. Lippert has many times more experience than me in the field of compilers, I like to make a few counter points:

1. On the point that it is silly:

Disagree, we also have the `??=` operator. Why? People could write `x = x ?? "";`, so is the `??=` also silly?

2. On the point that only a few people would actually use this feature:

I agree with the user "Ben Voight" in the comments on mr. Lippert's answer: 

> I've wanted short-circuiting compound assignment before. No one uses them because they don't exist.

This comment was upvoted 37 times (at time of writing), so clearly there are people that are interested in it. Because Stackoverflow is not a language design discussion page, the fact that this comment got so many upvotes indicates that there is in fact an audience for this feature.

3. On the point that you would be trading a test and branch, and thus make the job of the jitter and branch predictor more difficult:

In my above `IsUserAdmin` example, I prefer an extra branch in my code instead of always executing the "expensive" call if I would use the `|=` operator. (Which of course I know mr. Lippert would not suggest I do in this particular case). Furthermore, I think that the ability to write "expensive"/"bad" code should not be limited by the language.  **Simply put: the compiler should not hamper my ability to write bad code.**

**Comment by HaloFour** [comment, marked as answer]
See: #1718

**Reply by synercoder** [reply]
My search skills on this site have failed me...

**Reply by HaloFour** [reply]
No worries, a lot of issues/discussions can be hard to find, especially for operators based on symbols.  Only reason I knew to look was that I remembered it coming up before.  It doesn't help that the decision mentioned in that issue isn't recorded in the meeting notes.
